pkg/admission/prep_downscale.go

```diff
@@ -0,0 +1,211 @@
+		level.Debug(logger).Log("msg", "no replicas change, allowing")
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+	// Changes from/to nil scale are not downscales strictly speaking.
+	if oldReplicas == nil || newReplicas == nil {
+		return allowWarn(logger, "old/new replicas is nil, allowing the change")
+	}
+	// If it's not a downscale, just log debug.
+	if *oldReplicas < *newReplicas {
+		level.Debug(logger).Log("msg", "upscale allowed")
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+	if *oldReplicas == *newReplicas {
+		level.Debug(logger).Log("msg", "no replicas change, allowing")
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+
+	// Get the resource labels: for example, for a StatefulSet, it will be the labels of the StatefulSet itself,
+	// while for a Scale object, it will be the labels of the StatefulSet/Deployment/ReplicaSet that the Scale object belongs to.
+	var lbls map[string]string
+	switch o := oldObj.(type) {
+	case *appsv1.Deployment:
+		lbls = o.Labels
+	case *appsv1.StatefulSet:
+		lbls = o.Labels
+	case *appsv1.ReplicaSet:
+		lbls = o.Labels
+	case *autoscalingv1.Scale:
+		lbls, err = getResourceLabels(ctx, ar, api)
+		if err != nil {
+			return allowBecauseCannotGetResource(ar, logger, err)
+		}
+	default:
+		return allowWarn(logger, fmt.Sprintf("unsupported type %T, allowing the change", o))
+	}
+
+	if lbls[PrepDownscaleLabelKey] != PrepDownscaleLabelValue {
+		// Not labeled, nothing to do.
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+
+	port, ok := lbls[PrepDownscalePortKey]
+	if !ok || len(port) == 0 {
+		reviewResponse := v1.AdmissionResponse{
+			Allowed: false,
+			Result: &metav1.Status{
+				Message: fmt.Sprintf("downscale of %s/%s in %s from %d to %d replicas is not allowed because the %v label is not set or empty.", ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas, PrepDownscalePortKey),
+			},
+		}
+		level.Warn(logger).Log("msg", fmt.Sprintf("downscale not allowed because the %v label is not set or empty", PrepDownscalePortKey))
+		return &reviewResponse
+	}
+
+	path, ok := lbls[PrepDownscalePathKey]
+	if !ok || len(path) == 0 {
+		reviewResponse := v1.AdmissionResponse{
+			Allowed: false,
+			Result: &metav1.Status{
+				Message: fmt.Sprintf("downscale of %s/%s in %s from %d to %d replicas is not allowed because the %v label is not set or empty.", ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas, PrepDownscalePathKey),
+			},
+		}
+		level.Warn(logger).Log("msg", fmt.Sprintf("downscale not allowed because the %v label is not set or empty", PrepDownscalePathKey))
+		return &reviewResponse
+	}
+
+	// Since it's a downscale, check if the resource has the label that indicates it needs to be prepared to be downscaled.
+	// Create a slice of endpoint addresses for pods to send HTTP post requests to and to fail if any don't return 200
+	if !*ar.Request.DryRun {
```

Comment:
I'd invert this condition, and do an early return in case of dry run, to keep the indentation at minimum.